test/integration/protocol_integration_test.cc

```diff
@@ -2445,6 +2445,42 @@ TEST_P(DownstreamProtocolIntegrationTest, BasicMaxStreamTimeoutLegacy) {
   EXPECT_THAT(waitForAccessLog(access_log_name_), HasSubstr("max_duration_timeout"));
 }
 
+TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionNotReached) {
+  config_helper_.setDownstreamMaxStreamDuration(std::chrono::milliseconds(500));
+  config_helper_.setDownstreamMaxRequestsPerConnection(3);
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+
+  ASSERT_TRUE(response->waitForEndStream());
+  test_server_->waitForCounterEq("http.config_test.downstream_cx_max_requests", 0);
+}
+
+TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
+  config_helper_.setDownstreamMaxRequestsPerConnection(3);
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+
+  codec_client_->sendData(*request_encoder_, 1024 * 65, false);
+  codec_client_->sendData(*request_encoder_, 1024 * 65, false);
+  codec_client_->sendData(*request_encoder_, 1024 * 65, true);
+
+  ASSERT_TRUE(response->waitForEndStream());
```

Comment:
Getting 
```
Value of: response->waitForEndStream()
    Actual: false (Timed out waiting for end stream
  )
```

Reply:
You have a couple issues:

* You're only creating one stream, so you won't hit the limit as is. `startRequest` creates a new request; you'll need to do that up to the limit
* You're not responding to the request. Look at a test like `AddBodyToResponseAndWaitForIt` to see how to send a response back to complete a request. Once you've completed the first one, you can move to the next.

Reply:
I have updated the test.